src/util/bip143.rs

```diff
@@ -99,6 +99,117 @@ impl SighashComponents {
     }
 }
 
+/// A replacement for SigHashComponents which supports all sighash modes
+pub struct SigHashCache<'a> {
+    /// Access to transaction required for various introspection
+    tx: &'a Transaction,
+    /// Hash of all the previous outputs, computed as required
+    hash_prevouts: Option<sha256d::Hash>,
+    /// Hash of all the input sequence nos, computed as required
+    hash_sequence: Option<sha256d::Hash>,
+    /// Hash of all the outputs in this transaction, computed as required
+    hash_outputs: Option<sha256d::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &Transaction) -> SigHashCache {
+        SigHashCache {
+            tx: tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+        }
+    }
+
+    /// Calculate hash for prevouts
+    pub fn hash_prevouts(&mut self) -> sha256d::Hash {
+        if self.hash_prevouts.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txin in &self.tx.input {
+                txin.previous_output.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_prevouts = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_prevouts.unwrap()
+    }
+
+    /// Calculate hash for input sequence values
+    pub fn hash_sequence(&mut self) -> sha256d::Hash {
+        if self.hash_sequence.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txin in &self.tx.input {
+                txin.sequence.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_sequence = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_sequence.unwrap()
+    }
+
+    /// Calculate hash for outputs
+    pub fn hash_outputs(&mut self) -> sha256d::Hash {
+        if self.hash_outputs.is_none() {
+            let mut enc = sha256d::Hash::engine();
+            for txout in &self.tx.output {
+                txout.consensus_encode(&mut enc).unwrap();
+            }
+            self.hash_outputs = Some(sha256d::Hash::from_engine(enc));
+        }
+        self.hash_outputs.unwrap()
+    }
+
+    /// Compute the BIP143 sighash for any flag type. See SighashComponents::sighash_all simpler
+    /// API for the most common case
+    pub fn signature_hash(&mut self, input_index: usize, script_code: &Script, value: u64, sighash_u32: u32) -> SigHash {
```

Comment:
and as for keeping the same naming conventions all around the code, may be better to stick with `fn sig_hash` name?

Reply:
well just above it it's called `sighash_all` :) commence the bikeshedding!

Reply:
I do not think this is a bit issue with code completion; however for large libraries it's quite useful when you have a single naming rules which helps you guessing the function names w/o looking into the docs each time

Reply:
My point being it's currently two things:

`sighash_all` and `signature_hash`. So making it `sig_hash` would not make things any more cohesive?

Reply:
If we are getting rid of SigHash type, then no. I think `sighash_*` in this regard is prefered over `signature_hash[_*]` for sure.